### ISSUE TITLE
Something wrong with cdc.UnmarshalBinaryBare?

### ISSUE THREAD

**Issue by james-ray** [tendermint/tendermint#3236]
**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source):

**ABCI app** (name for built-in, URL for self-written if it's publicly available):

**Environment**:
- **OS** (e.g. from /etc/os-release):
- **Install tools**:
- **Others**:

**What happened**:
I was trying to write a test case, but found mockApp trigger a nil-pointer exception
func (mock *mockProxyApp) DeliverTx(tx []byte) abci.ResponseDeliverTx {
	r := mock.abciResponses.DeliverTx[mock.txCount]
	mock.txCount++
	return *r   //this line nil-pointer
}
then I found it indeed is a nil pointer, then I add some code in ApplyBlock:

	bytes:=cdc.MustMarshalBinaryBare(abciResponses)
	newabciResponses := new(ABCIResponses)
	err = cdc.UnmarshalBinaryBare(bytes, newabciResponses)
	if err != nil {
		// DATA HAS BEEN CORRUPTED OR THE SPEC HAS CHANGED
		fmt.Sprintf(`LoadABCIResponses: Data has been corrupted or its spec has
                changed: %v\n`, err)
	}

	// Save the results before we commit.
	saveABCIResponses(blockExec.db, block.Height, abciResponses)

I set a breakpoint after UnmarshalBinaryBare and found the newabciResponses contains nil DeliverTxResponse, while abciResponse contains not nil DeliverTxResponse. The fields in DeliverTxResponse returned is all 0 or empty bytes, but we should not lose the structure, Unmarshal return a nil-pointer.

**What you expected to happen**:

**Have you tried the latest version**: yes/no

**How to reproduce it** (as minimally and precisely as possible):

**Logs (paste a small part showing an error (< 10 lines) or link a pastebin, gist, etc. containing more of the log file)**:

**Config (you can paste only the changes you've made)**:

**node command runtime flags**:

**`/dump_consensus_state` output for consensus bugs**

**Anything else we need to know**:

**Comment by Liamsi** [comment]
I'll check if this is an amino issue or a problem on how this is used.

**Comment by james-ray** [comment]
I write a very simple test:

func TestWwwww(t *testing.T) {
	bytes,_:=hex.DecodeString("0A0012310A2F0A2B0A0765643235353139122076F5D0FACA2A5BD1825E2B554B03AFBD06D305C6BCF22ACE163776B01FC7682D100A1A00D")
	newabciResponses := new(state.ABCIResponses)
	err := cdc.UnmarshalBinaryBare(bytes, newabciResponses)
	if err != nil {
		// DATA HAS BEEN CORRUPTED OR THE SPEC HAS CHANGED
		fmt.Sprintf(`LoadABCIResponses: Data has been corrupted or its spec has
            changed: %v\n`, err)
	}
	fmt.Printf("%v",newabciResponses)
}

=== RUN   TestWwwww
&{[<nil>] validator_updates:<pub_key:<type:"ed25519" data:"v\365\320\372\312*[\321\202^+UK\003\257\275\006\323\005\306\274\362*\316\0267v\260\037\307h-" > power:10 >  }--- PASS: TestWwwww (581.70s)
PASS

**Comment by james-ray** [comment]
another 2 tests:
func TestWwwww1(t *testing.T) {
	abciResponses := new(state.ABCIResponses)
	abciResponses.DeliverTx=make([]*abci.ResponseDeliverTx,0)
	abciResponses.DeliverTx=append(abciResponses.DeliverTx,&abci.ResponseDeliverTx{})
	bytes := cdc.MustMarshalBinaryBare(abciResponses)
	fmt.Printf("marshal bytes: %X", bytes)
	fmt.Println()
	newabciResponses := new(state.ABCIResponses)
	err := cdc.UnmarshalBinaryBare(bytes, newabciResponses)
	if err != nil {
		// DATA HAS BEEN CORRUPTED OR THE SPEC HAS CHANGED
		fmt.Sprintf(`LoadABCIResponses: Data has been corrupted or its spec has
            changed: %v\n`, err)
	}
	fmt.Printf("unmarshal: %v", newabciResponses)
}

=== RUN   TestWwwww1
marshal bytes: 0A00
unmarshal: &{[<nil>] <nil> <nil>}--- PASS: TestWwwww1 (0.00s)
PASS



func TestWwwww2(t *testing.T) {
	abciResponses := new(state.ABCIResponses)
	abciResponses.DeliverTx=make([]*abci.ResponseDeliverTx,0)
	abciResponses.DeliverTx=append(abciResponses.DeliverTx,&abci.ResponseDeliverTx{Info:"hello"})
	bytes := cdc.MustMarshalBinaryBare(abciResponses)
	fmt.Printf("marshal bytes: %X", bytes)
	fmt.Println()
	newabciResponses := new(state.ABCIResponses)
	err := cdc.UnmarshalBinaryBare(bytes, newabciResponses)
	if err != nil {
		// DATA HAS BEEN CORRUPTED OR THE SPEC HAS CHANGED
		fmt.Sprintf(`LoadABCIResponses: Data has been corrupted or its spec has
            changed: %v\n`, err)
	}
	fmt.Printf("unmarshal: %v", newabciResponses)
}


=== RUN   TestWwwww2
marshal bytes: 0A07220568656C6C6F
unmarshal: &{[info:"hello" ] <nil> <nil>}--- PASS: TestWwwww2 (0.00s)
PASS

**Comment by Liamsi** [comment]
Thanks a lot @james-ray.
 
In amino, like in protobuf, there is no way to differentiate between "empty" or "default" (like 0 for instance); i.e. both will be encoded the same. (There are minor differences on how proto and amino treat those but those should not matter here).

I'm just trying to understand if this is an actual problem or not (is there any place where the it actually matters?). 

I slightly modified your test to explain this further:

```go
func TestAminoAbciResponsesRountrip(t *testing.T) {
	abciResponses := new(state.ABCIResponses)
	abciResponses.DeliverTx=make([]*abci.ResponseDeliverTx,0)
	abciResponses.DeliverTx=append(abciResponses.DeliverTx,&abci.ResponseDeliverTx{})
	bytes := cdc.MustMarshalBinaryBare(abciResponses)
	newabciResponses := new(state.ABCIResponses)
	err := cdc.UnmarshalBinaryBare(bytes, newabciResponses)
	require.NoError(t, err)

        // This will fail:
	assert.Equal(t, abciResponses, newabciResponses)
}
```

In other words, does tendermint contain any lines of code that e.g. would rely on  the field `DeliverTx` being not nil but a pointer to an empty struct instead? The only place I can find is `mockProxyApp`. It is passed the abciResponses returned by `LoadABCIResponses` which calls `UnmarshalBinaryBare`. I think the `mockProxyApp` should test for `nil` where it makes sense (in `DeliverTx` and `EndBlock`)! Or am I missing sth here?

**Comment by melekes** [comment]
@james-ray any thoughts on ^^^?

**Comment by james-ray** [comment]
@Liamsi 
I run your last test, but it turns out panic

=== RUN   TestMockProxyApp
--- FAIL: TestMockProxyApp (0.00s)
	reactor_test.go:427: 
			Error Trace:	reactor_test.go:427
			Error:      	func (assert.PanicTestFunc)(0xe2dd80) should not panic
			            		Panic value:	runtime error: invalid memory address or nil pointer dereference
			Test:       	TestMockProxyApp
FAIL

Yes, I think it is the problem I'm discussing.

In state_test, we will test kvStore, when it's a updateValidatorTx, it will return an empty abci.ResponseDeliverTx

**Comment by james-ray** [comment]
So I think either we ensure our every code don't contain an empty structure or judge nil pointer before we use ?

**Comment by james-ray** [comment]
Since replay blocks containing validators change tx is truely possible, and kvStore is a sample app. I think maybe it is indeed a problem how we use amino. Then we should notify developers to avoid return wholly empty ResponseDeliverTx.

**Comment by james-ray** [comment]
the panic code is in MockApp

func (mock *mockProxyApp) DeliverTx(tx []byte) abci.ResponseDeliverTx {
	r := mock.abciResponses.DeliverTx[mock.txCount]
	mock.txCount++
	return *r
}

or else we could judge the r here , if it is nil, we return an empty ResponseDeliverTx.

**Comment by james-ray** [comment]
Allow me to do it.  I can fix it in #3067 just by the way.

**Comment by Liamsi** [comment]
Cool, Thanks!

**Comment by marbar3778** [comment]
This issue can be closed as now there is a safety check on save to remove nil values. I ran into this when migrating to protobuf but now it should be fixed
